### PR TITLE
mod 9: use mongodb

### PR DESCRIPTION
**Pull request by JoelWee**
For some reason running pytest directly doesn't work. it looks like it's not resetting the DB that's being connected to (e.g. my e2e tests look like they're reading from the mockupDB still)

### CODE REVIEW

File: app/flask_config.py [hunk 1]
```diff
@@ -5,11 +5,7 @@
...
 
-    TRELLO_API_TOKEN = os.environ.get("TRELLO_API_TOKEN")
-    TRELLO_API_KEY = os.environ.get("TRELLO_API_KEY")
-    TRELLO_BOARD_ID = os.environ.get("TRELLO_BOARD_ID")
-
-    if not (TRELLO_API_KEY and TRELLO_API_TOKEN and TRELLO_BOARD_ID):
-        raise ValueError(
-            "Missing config for Flask application. Did you forget to run setup.sh?"
-        )
+    MONGO_URI = os.environ.get("MONGO_URI")
```
Comment (hugh-emerson) [inline app/flask_config.py:8]: I assume your comment about being unable to run the tests directly you mean `poetry run pytest` doesn't pass. This is why you are not able to run all the tests together - this line setting MONGO_URI is only evaluated once ever (the first time your app is created)

You can still successfully run `poetry run pytest tests`/`poetry run pytest tests_e2e` separately.
If you want to run them all together, you will need to change this to an instance attribute and change the line in `app/__init__.py` to something like `app.config.from_object(Config())`

File: pyproject.toml [hunk 2]
```diff
@@ -8,17 +8,25 @@ authors = ["Joel Wee <joel.wee@outlook.com>"]
...
-requests = "^2.24.0"
-flake8 = "^3.8.3"
-watchdog = "^0.10.3"
+mockupdb = "^1.8.0"
+watchdog = { version = "^0.10.3", optional = true }
+pyyaml = { version = "^5.3.1", optional = true }
+rope = { version = "^0.17.0", optional = true }
+flake8 = { version = "^3.8.3", optional = true }
+autopep8 = { version = "^1.5.5", optional = true }
+black = { version = "^20.8b1", optional = true }
```
Comment (hugh-emerson) [inline pyproject.toml:25]: Are these really necessary?

File: docker-compose.test.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      target: ci
+      dockerfile: Dockerfile.test
+    environment:
+      - "MONGO_TEST_BASE_URI=mongodb://db:27017"
+    depends_on:
+      - db
+  db:
+    image: mongo
+    ports:
+      - "27017:27017"
```
Comment (hugh-emerson) [inline docker-compose.test.yml]: This isn't necessary unless you want to try accessing it directly from your host machine. The mongo image already "exposes" 27017.

File: tests_e2e/test_app.py [hunk 4]
```diff
@@ -22,13 +22,9 @@ def driver():
...
-    load_dotenv(".env", override=True)
-    trello_board = TrelloBoard(
-        key=os.environ["TRELLO_API_KEY"], token=os.environ["TRELLO_API_TOKEN"]
-    )
-
-    board_id = trello_board.create("test")
-    os.environ["TRELLO_BOARD_ID"] = board_id
+    mongo_test_base_uri = os.environ.get("MONGO_TEST_BASE_URI", default_mongo_uri)
+    mongo_db = "test"
+    os.environ["MONGO_URI"] = f"{mongo_test_base_uri}/{mongo_db}"
```
Comment (hugh-emerson) [inline tests_e2e/test_app.py]: Why do this in two steps with separate MONGO_TEST_BASE_URI and MONGO_URI env vars?

File: docker-compose.test.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  tests:
+    build:
+      context: .
+      target: ci
+      dockerfile: Dockerfile.test
+    environment:
+      - "MONGO_TEST_BASE_URI=mongodb://db:27017"
+    depends_on:
+      - db
+  db:
```
Comment (hugh-emerson) [inline docker-compose.test.yml]: Potentially nicer to use a separate name for this service compared to the normal compose file to avoid conflicts between the two. But then again, you do use a named "test" database for the tests so it shouldn't be an issue.

File: pyproject.toml [hunk 6]
```diff
@@ -8,17 +8,25 @@ authors = ["Joel Wee <joel.wee@outlook.com>"]
 python = "^3.8"
 flask = "^1.1.2"
 python-dotenv = "^0.14.0"
-requests_mock = "^1.8.0"
 gunicorn = "^20.0.4"
-pyyaml = "^5.3.1"
+pymongo = "^3.11.3"
+dnspython = "^2.1.0"
```
Comment (hugh-emerson) [inline pyproject.toml:13]: Why is this necessary?

File: tests/mock_data/items.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+json = [
```
Comment (hugh-emerson) [inline tests/mock_data/items.py]: Slightly confusing for this to be called `json` when it's just python code, not json.

File: README.md [hunk 8]
```diff
@@ -7,6 +7,7 @@ The project uses a virtual environment to isolate package dependencies.  You wil
 ### On Windows (Using Git Bash), macOS and Linux
 ```bash
 cp -n .env.template .env
+poetry install -E linting
 ```
 
 Once the setup script has completed and all packages have been installed, start the Flask app by running:
```
Comment (hugh-emerson) [inline README.md:13]: You haven't mentioned setting up MONGO_URI in .env.

File: .travis.yml [hunk 9]
```diff
@@ -1,29 +1,36 @@
...
-  - echo $DOCKER_HUB_PASSWORD | docker login -u j0elwee --password-stdin registry.heroku.com
+  - echo $HEROKU_API_KEY | docker login -u j0elwee --password-stdin registry.heroku.com
   - docker tag j0elwee/devops-trg registry.heroku.com/devops-trg/web
   - docker push registry.heroku.com/devops-trg/web
 deploy:
   provider: script
   script: heroku container:release web -a devops-trg
   on: 
-    branch: master
+    branch: mod-9-db
```
Comment (hugh-emerson) [inline .travis.yml]: Don't forget to revert this.

**Review by hugh-emerson — APPROVED** [review]
Looks good! See my comments for some minor suggestions, but merge when you're ready.

File: pyproject.toml (hunk 6 above)
Comment (JoelWee) [reply pyproject.toml:13]: It complains about this dependency if I don't include it
